daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+ * already exist at the given path, a new one is created. Limits the physical
+ * size of the underlying data file to max_size bytes. If a circular file with a
+ * different maximum size already exists at the given path, its maximum size is
+ * changed to the given value, which may result in data loss.
+ */
+EmerCircularFile *
+emer_circular_file_new (const gchar *path,
+                        guint64      max_size,
+                        GError     **error)
+{
+  return g_initable_new (EMER_TYPE_CIRCULAR_FILE,
+                         NULL /* GCancellable */,
+                         error,
+                         "path", path,
+                         "max-size", max_size,
+                         NULL);
+}
+
+/* Appends the given element in-memory only. Use emer_circular_file_save to
+ * flush all appended elements. This allows for batching of writes. Note that
+ * elements can not be read with emer_circular_file_read until they have been
+ * saved. Returns TRUE if the given element was successfully appended and will
+ * fit in the space allotted to the circular file. Returns FALSE otherwise.
+ */
+gboolean
+emer_circular_file_append (EmerCircularFile *self,
+                           gconstpointer     elem,
+                           guint64           elem_size)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  guint64 elem_size_on_disk = sizeof (elem_size) + elem_size;
+  guint64 total_size = priv->size + priv->write_buffer->len + elem_size_on_disk;
+  if (total_size > priv->max_size)
+    return FALSE;
+
+  guint64 little_endian_elem_size = swap_bytes_64_if_big_endian (elem_size);
+  g_byte_array_append (priv->write_buffer,
+                       (const guint8 *) &little_endian_elem_size,
+                       sizeof (little_endian_elem_size));
+  g_byte_array_append (priv->write_buffer, elem, elem_size);
+
+  return TRUE;
+}
+
+/* Flushes all elements successfully appended via emer_circular_file_append
+ * through to the underlying data file. Elements are saved in the same order in
+ * which they were appended. Returns TRUE on success and FALSE on error.
+ */
+gboolean
+emer_circular_file_save (EmerCircularFile *self,
+                         GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  if (priv->write_buffer->len == 0)
+    return TRUE;
+
+  GFileOutputStream *file_output_stream =
+    g_file_append_to (priv->data_file, G_FILE_CREATE_NONE,
+                      NULL /* GCancellable */, error);
+  if (file_output_stream == NULL)
+    return FALSE;
+
+  GOutputStream *output_stream = G_OUTPUT_STREAM (file_output_stream);
+  GSeekable *seekable = G_SEEKABLE (file_output_stream);
+
+  goffset tail = (priv->head + priv->size) % priv->max_size;
+  guint64 space_available_at_tail = priv->max_size - tail;
+  gsize bytes_tail;
+  if (priv->write_buffer->len <= space_available_at_tail)
```

Comment:
`gssize` is only guaranteed to go down to `-1` (http://pubs.opengroup.org/onlinepubs/009696699/basedefs/sys/types.h.html), but I suspect it would work in practice.